hngprojects/hng_boilerplate_nestjs
# [FEAT] Implement Backend for Notification Settings

# What does this PR do?
This PR implements endpoints for managing user notification settings:
* Added POST /api/v1/settings/notification-settings Endpoint: Creates or updates user notification settings.
* Added GET /api/v1/settings/notification-settings/:user_id Endpoint: Retrieves user notification settings.
* Authentication: Ensures only authenticated users can access these endpoints.
* Validation: Implements input validation for notification settings data.
* Error Handling: Returns specific HTTP status codes and messages based on various scenarios.

# How should this be manually tested?
1. Login User:
   * Make a `POST` request to `/api/v1/auth/login` with a valid user's email and password.
   * Save the access token from the response.

2. Create/Update Notification Settings:
   * Make a `POST` request to `/api/v1/settings/notification-settings` with the following:
     - Header: `Authorization: Bearer <access_token>`
     - Body:
       ```json
       {
         "user_id": "user@example.com",
         "email_notifications": true,
         "push_notifications": false,
         "sms_notifications": true
       }
       ```
   * Verify the response contains the created/updated notification settings.

3. **Get Notification Settings:**
   * Make a `GET` request to `/api/v1/settings/notification-settings/user@example.com` with the access token.
   * Verify the response contains the user's notification settings.

4. Edge Cases:
   * Unauthenticated: Test without a token to ensure the endpoint returns a `401 Unauthorized` status.
   * Invalid User: Test with a non-existent user_id to ensure the endpoint returns a `404 Not Found` status.
   * Invalid Input: Test with invalid data to ensure proper validation errors are returned.

# Checklist:
- [x] Implemented `POST /api/v1/settings/notification-settings` endpoint in `notification-settings.controller.ts`.
- [x] Implemented `GET /api/v1/settings/notification-settings/:user_id` endpoint in `notification-settings.controller.ts`.
- [x] Added `createOrUpdate` and `findByUserId` methods in the `NotificationSettingsService`.
- [x] Implemented input validation using DTO.
- [x] Tested endpoint functionality to ensure it correctly handles various scenarios.
- [x] Added unit tests for the `NotificationSettingsService`.
- [x] Updated API documentation to reflect the new endpoints.

# Response Examples
**Success (POST/GET):**

```json
{
  "status": "success",
  "data": {
    "id": "a9f4b3b0-4b3b-4b3b-4b3b-4b3b4b3b4b3b",
    "user_id": "user@example.com",
    "email_notifications": true,
    "push_notifications": false,
    "sms_notifications": true,
    "created_at": "2024-07-25T12:00:00Z",
    "updated_at": "2024-07-25T12:00:00Z"
  }
}
```

**Error (Unauthorized):**

```json
{
  "message": "Unauthorized",
  "status_code": 401
}
```

**Error (Not Found):**

```json
{
  "status": "error",
  "message": "Notification settings not found",
  "status_code": 404
}
```

**Error (Validation):**

```json
{
  "status": "error",
  "message": [
    "user_id must be a valid email",
    "email_notifications must be a boolean",
    "push_notifications must be a boolean",
    "sms_notifications must be a boolean"
  ],
  "status_code": 400
}
```

# Link to Issue
Link to this issue: [#[FEAT] Implement Backend for Notification Settings #20](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/20)
